core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1872,9 +1880,9 @@ class KafkaController(val config: KafkaConfig,
     } else {
       val reassignmentResults = mutable.Map.empty[TopicPartition, ApiError]
       val partitionsToReassign = mutable.Map.empty[TopicPartition, ReplicaAssignment]
-
+      val noNewPartitionBrokers = partitionUnassignableBrokerIds.toSet
       reassignments.foreach { case (tp, targetReplicas) =>
-        if (replicasAreValid(tp, targetReplicas)) {
+        if (replicasAreValid(tp, targetReplicas, noNewPartitionBrokers)) {
```

Comment:
still have some race condition here,  because a broker can be marked as maintenance brokers after partition reassignment request is received before the reassignment request is completed